acb_dirichlet/backlund_s.c

```diff
@@ -0,0 +1,71 @@
+/*
+    Copyright (C) 2019 D.H.J. Polymath
+
+    This file is part of Arb.
+
+    Arb is free software: you can redistribute it and/or modify it under
+    the terms of the GNU Lesser General Public License (LGPL) as published
+    by the Free Software Foundation; either version 2.1 of the License, or
+    (at your option) any later version.  See <http://www.gnu.org/licenses/>.
+*/
+
+#include "acb_dirichlet.h"
+
+void
+acb_dirichlet_backlund_s(arb_t res, const arb_t t, slong prec)
+{
+    mag_t m;
+
+    if (!arb_is_nonnegative(t))
+    {
+        arb_indeterminate(res);
+        return;
+    }
+
+    mag_init(m);
+
+    /*
+     * If the error radius of t is greater than 8/log(t) then use a fast bound.
+     * The threshold doesn't matter for correctness, but note that the average
+     * spacing of zeros is 2*pi/log(t) asymptotically.
+     * Otherwise if the radius is smaller use S(t) = N(t) - theta(t)/pi - 1.
+     */
+    arb_get_mag(m, t);
+    mag_log(m, m);
+    mag_mul_2exp_si(m, m, -3);
+    mag_inv(m, m);
+    if (mag_cmp(arb_radref(t), m) > 0)
+    {
+        arb_zero(res);
+        acb_dirichlet_backlund_s_bound(arb_radref(res), t);
+    }
+    else
+    {
+        arb_t N;
+        acb_t z;
+
+        arb_init(N);
+        acb_init(z);
+
+        /*
+         * The difference (N(t) - theta(t)/pi) loses precision from
+         * cancellation. Add extra bits to mitigate the loss.
+         */
+        acb_set_arb(z, t);
+        acb_dirichlet_hardy_theta(z, z, NULL, NULL, 1, prec);
+        prec += acb_bits(z);
```

Comment:
I also saw the use of arf_bits in the zeta zeros code, which is dubious. The number of bits doesn't necessarily correspond to the precision, accuracy or magnitude; in fact, it will fluctuate depending on the number of trailing zeros created by rounding. Shouldn't the precision just be doubled here?

Reply:
Alternatively, incremented by the log2 of the magnitude of t...

Reply:
Thanks for the correction, I've changed both instances to now use log2 of magnitude.

Reply:
There was a merge conflict that I've tried to resolve by undoing my changes to the zeta zeros code.